### ISSUE TITLE
False positive on match_same_arms

### ISSUE THREAD

**Issue by llogiq** [rust-lang/rust-clippy#860]
Example: Any kind of `xor`-like handling, e.g. let `Foo` be one variant of an enum:

```
match (x, y) {
    (Foo(_), Foo(_)) => None,
    (Foo(x), _) |
    (_, Foo(x)) => Some(x),
    _ => None
}
```

This will trigger the lint, because the _position_ of the arms is not taken into account.

@regexident suggested to only lint adjacent same arms. Another option would be to check if those arms are disjunct and can be reordered without changing semantics, but that is obviously much much harder to do in the general case.

**Comment by Boddlnagg** [comment]
This lint also complained about this `match`, what I would consider to be a false positive:
```rust
match *self {
    SendError::InvalidData(msg) => msg,
    SendError::Other(msg) => msg
}
````

**Comment by mcarton** [comment]
Why?
This can be written

```rust
match *self {
    SendError::InvalidData(msg) | SendError::Other(msg) => msg
}
```

**Comment by Boddlnagg** [comment]
Oh, right. I've never used `|` before, so I totally didn't think about it that way. Maybe the documentation could mention it, in case there are other people out there who don't think about this ;-)

**Comment by mcarton** [comment]
[The documentation *does* mention it :smile:](https://github.com/Manishearth/rust-clippy/wiki#match_same_arms)

**Comment by Boddlnagg** [comment]
Now _this_ is really embarrassing, sorry. I wanted to report another clippy issue right now, but I'm going to sleep on it first.

**Comment by Manishearth** [comment]
Eh, report away, false positives aren't a big deal.

And in this case we should be suggesting the |. Documentation is fine for extended things,  but suggestions should be prominent :smile:

**Comment by Manishearth** [comment]
(when I say false positives I mean false positives when filing bugs, not clippy false positives )

**Comment by liamcurry** [comment]
Might also be worth mentioning matching on enums with named fields of the same name, e.g.:

```rust
pub enum Test {
    A { field: String },
    B { field: String },
}

impl Test {
    /// `match_same_arms` warning
    fn field_bad(&self) -> &str {
        match self {
            Self::A { field } => field,
            Self::B { field } => field,
        }
    }
    /// no warning
    fn field_good(&self) -> &str {
        match self {
            Self::A { field } | Self::B { field } => field,
        }
    }
}
```
